### PR TITLE
Prototype of particle system

### PR DESCRIPTION
**Pull request by celmer89**
first iteration of particle system

### CODE REVIEW

**Comment by keepermanj** [comment]
Hi Celmer,

Good to see another member of the community on here, playing with code. How would I see the particle system in action?

File: animation.cpp [hunk 1]
```diff
@@ -72,3 +72,94 @@ PAnimation Animation::fromId(AnimationId id) {
         {{577, 598}, {94, 94}, {-29, -29}},
         }));
 }
+
+
+class Particle{
```
Comment (miki151) [inline animation.cpp]: Always put one space before {

File: animation.cpp [hunk 2]
```diff
@@ -72,3 +72,94 @@ PAnimation Animation::fromId(AnimationId id) {
...
+  
+  public:
+    Particle(Vec2 position, Vec2 origin, Color color, Vec2 speed, int size):
+     _position(position), _origin(origin), _color(color), _speed(speed), _size(size)
+    {
```
Comment (miki151) [inline animation.cpp]: Put { at the end of previous line

File: animation.cpp [hunk 3]
```diff
@@ -72,3 +72,94 @@ PAnimation Animation::fromId(AnimationId id) {
         {{577, 598}, {94, 94}, {-29, -29}},
         }));
 }
+
+
+class Particle{
+  
+  public:
+    Particle(Vec2 position, Vec2 origin, Color color, Vec2 speed, int size):
+     _position(position), _origin(origin), _color(color), _speed(speed), _size(size)
```
Comment (miki151) [inline animation.cpp]: Put 4 spaces of extra indent when breaking line, also put : in the next line.

File: animation.cpp [hunk 4]
```diff
@@ -72,3 +72,94 @@ PAnimation Animation::fromId(AnimationId id) {
...
+    {
+
+      _color.a = 128;
+    }
+
+    void Update(){
+
```
Comment (miki151) [inline animation.cpp]: Remove these extra lines please.

File: animation.cpp [hunk 5]
```diff
@@ -72,3 +72,94 @@ PAnimation Animation::fromId(AnimationId id) {
...
+      // custom prticle effect logic
+       _position +=_speed;
```
Comment (miki151) [inline animation.cpp]: Add extra space around operators like += / etc.

File: animation.cpp [hunk 6]
```diff
@@ -72,3 +72,94 @@ PAnimation Animation::fromId(AnimationId id) {
...
+      _color.a-=10;
+       
+    }
+
+    void Draw(Renderer& renderer){
+
+      renderer.drawPoint(_position + _origin, _color, _size);
+    }
+
+  private:
```
Comment (miki151) [inline animation.cpp]: private: and public: should have the same indentation as the rest of the class.

File: animation.cpp [hunk 7]
```diff
@@ -72,3 +72,94 @@ PAnimation Animation::fromId(AnimationId id) {
...
+    
+    Vec2 _position;
+    Vec2 _origin;
+    Color  _color;
```
Comment (miki151) [inline animation.cpp]: Remove extra space

File: animation.cpp [hunk 8]
```diff
@@ -72,3 +72,94 @@ PAnimation Animation::fromId(AnimationId id) {
...
+
+
+
+class ParticleEffect : public Animation {
+
+  public:
+  ParticleEffect(milliseconds duration, unsigned int particleNum, Vec2 origin)
+      : Animation(duration), _particleNum(particleNum), _origin(origin)  {
+
+        // custom particel effect setup
```
Comment (miki151) [inline animation.cpp]: Indentation should be 2 spaces to the right of function declaration.

File: animation.cpp [hunk 9]
```diff
@@ -72,3 +72,94 @@ PAnimation Animation::fromId(AnimationId id) {
...
+        _particleNum = 8;
+        _particles.reserve(_particleNum);
+        int size = 15;
+        
+        _particles.push_back(Particle(Vec2(1,1), _origin, Color::LIGHT_RED,Vec2(3,3), size));
```
Comment (miki151) [inline animation.cpp]: Please always add a space after ,

File: animation.cpp [hunk 10]
```diff
@@ -72,3 +72,94 @@ PAnimation Animation::fromId(AnimationId id) {
...
+  
+  public:
+    Particle(Vec2 position, Vec2 origin, Color color, Vec2 speed, int size):
+     _position(position), _origin(origin), _color(color), _speed(speed), _size(size)
+    {
+
+      _color.a = 128;
+    }
+
+    void Update(){
```
Comment (miki151) [inline animation.cpp]: Method names should start with lowercase letter

File: animation.cpp [hunk 11]
```diff
@@ -72,3 +72,94 @@ PAnimation Animation::fromId(AnimationId id) {
...
+    }
+    // render particles
+    for (auto& particle : _particles){
+      particle.Draw(renderer);
+    } 
+  } 
+
+  private:
+
+    unsigned int _particleNum;
```
Comment (miki151) [inline animation.cpp]: Remove _ from member names.

File: renderer.cpp [hunk 12]
```diff
@@ -437,6 +437,15 @@ void Renderer::drawFilledRectangle(int px, int py, int kx, int ky, Color color,
   drawFilledRectangle(Rectangle(px, py, kx, ky), color, outline);
 }
 
+void Renderer::drawPoint(Vec2 pos, Color color, int size) {
+
+ SDL::  glPointSize(size);
+ SDL::glBegin(GL_POINTS);
+    color.applyGl();
+		SDL::glVertex2f(pos.x, pos.y );
```
Comment (miki151) [inline renderer.cpp]: Please never use tabs in the code.

File: renderer.cpp [hunk 13]
```diff
@@ -437,6 +437,15 @@ void Renderer::drawFilledRectangle(int px, int py, int kx, int ky, Color color,
   drawFilledRectangle(Rectangle(px, py, kx, ky), color, outline);
 }
 
+void Renderer::drawPoint(Vec2 pos, Color color, int size) {
+
+ SDL::  glPointSize(size);
```
Comment (miki151) [inline renderer.cpp]: Indentation should be 2 spaces. Remove extra spaces after ::

File: window_view.cpp [hunk 14]
```diff
@@ -1458,6 +1458,16 @@ void WindowView::keyboardAction(const SDL_Keysym& key) {
       inputQueue.push(UserInput(getDirActionId(key), Vec2(-1, -1)));
       mapGui->onMouseGone();
       break;
+   case SDL::SDLK_TAB:
```
Comment (miki151) [inline window_view.cpp]: Fix indentation please.

File: animation.cpp [hunk 15]
```diff
@@ -72,3 +72,78 @@ PAnimation Animation::fromId(AnimationId id) {
...
+class Particle {
+  public:
+  Particle(Vec2 position, Vec2 origin, Color color, Vec2 speed, int size)
+      : position(position), origin(origin), color(color), speed(speed), size(size) {
+    color.a = 128;
+  }
+
+  void update() {
+    // custom particle effect logic
+    position +=speed;
```
Comment (miki151) [inline animation.cpp]: Please add space between += and speed

File: animation.cpp [hunk 16]
```diff
@@ -72,3 +72,78 @@ PAnimation Animation::fromId(AnimationId id) {
...
+
+  private:
+  Vec2 position;
+  Vec2 origin;
+  Color color;
+  Vec2 speed;
+  int size;
+
+  friend class ParticleEffect;
+  };
```
Comment (miki151) [inline animation.cpp]: Bad indent of };

File: animation.cpp [hunk 17]
```diff
@@ -72,3 +72,78 @@ PAnimation Animation::fromId(AnimationId id) {
...
+
+class ParticleEffect : public Animation {
+  public:
+  ParticleEffect(milliseconds duration, unsigned int particleNum, Vec2 origin)
```
Comment (miki151) [inline animation.cpp]: It seems that the particleNum parameter is unnecessary. It's overwritten to 8 just below, and actually never read.

File: animation.cpp [hunk 18]
```diff
@@ -72,3 +72,78 @@ PAnimation Animation::fromId(AnimationId id) {
...
+    } 
+  } 
+
+  private:
+  unsigned int particleNum;
+  Vec2 origin;
+  vector<Particle> particles;
+};
+
+PAnimation Animation::perticleEffect(int id, milliseconds duration, unsigned int particleNum, Vec2 origin){
```
Comment (miki151) [inline animation.cpp]: Add space between ) and {

File: renderer.cpp [hunk 19]
```diff
@@ -437,6 +437,14 @@ void Renderer::drawFilledRectangle(int px, int py, int kx, int ky, Color color,
   drawFilledRectangle(Rectangle(px, py, kx, ky), color, outline);
 }
 
+void Renderer::drawPoint(Vec2 pos, Color color, int size) {
+  SDL::glPointSize(size);
+  SDL::glBegin(GL_POINTS);
+  color.applyGl();
+  SDL::glVertex2f(pos.x, pos.y );
```
Comment (miki151) [inline renderer.cpp]: Remove space before );

File: window_view.cpp [hunk 20]
```diff
@@ -1458,6 +1458,18 @@ void WindowView::keyboardAction(const SDL_Keysym& key) {
       inputQueue.push(UserInput(getDirActionId(key), Vec2(-1, -1)));
       mapGui->onMouseGone();
       break;
+  #ifndef RELEASE
+    case SDL::SDLK_TAB:
```
Comment (miki151) [inline window_view.cpp]: Please move the whole case statement above so we don't duplicate the #ifndef RELEASE. Also { should be moved to the line above.

**Pull request merged**